### PR TITLE
Draft: ZPT-calculations

### PR DESCRIPTION
**Pull request by joakimmoller**
Finally!

Draft pullrequest for refactoring Donaletty to make zpt-profiles for scans. This is merely a proof of concept.

Following stuff not implemented yet:
- [ ] stack
- [ ] step functions
- [ ] tests other than manual (mytest.py)
- [ ] Scientific review

### CODE REVIEW

File: app/zpt/handler/atmos.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import numpy as np
+from scipy.integrate import odeint  # type: ignore
+from scipy.interpolate import BSpline, splrep  # type: ignore
+
+from .geos import g  # type:ignore
+
+AVOGADRO = 6.02282e23  # [mol^-1] aovogadros number
```
Comment (skymandr) [inline app/zpt/handler/atmos.py:7]:
```suggestion
AVOGADRO = 6.02214076e23   # [mol^-1] aovogadros number
```
This actually has an exact definition these days, so no reason not to use it: https://www.nist.gov/si-redefinition/meet-constants

File: app/zpt/handler/atmos.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+
+def intermbar(z):
+    mbars = np.r_[
```
Comment (skymandr) [inline app/zpt/handler/atmos.py:11]: `np.r_[...]` in this case only does the exact same thing `np.array([...])` does. I think the latter looks a little less obscure. It also seems a bit wasteful to reinitialize `mbars` and `mbarz` every time the function is called, but maybe it is only called once?

File: app/zpt/handler/atmos.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+def intatm(z, T, newz, normz, normrho, lat):
+    """Integrates the temperature profile to yield a new model
+    atmosphere in hydrostatic equilibrium including the effect of
+    g(z) and M(z).
+    newT, p, rho, nodens, n2, o2, o = intatm(z, T, newz, normz,normrho)
+    NOTE z in km and returns pressure in pa
+    """
+    wn2 = 0.78084  # mixing ratio N2
+    wo2 = 0.209476  # mixing ratio O2
+    k = 1.38054e-23  # jK-1 Boltzmans constant
```
Comment (skymandr) [inline app/zpt/handler/atmos.py:41]:
```suggestion
    k = 1.380649e-23 # jK-1 Boltzmans constant
```
This also has an exact definition: https://www.nist.gov/si-redefinition/meet-constants

File: app/zpt/handler/geoloc_tools.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+def cart2sph(x_cart, y_cart, z_cart):
+    hypot_xy = N.hypot(x_cart, y_cart)
+    radius = N.hypot(hypot_xy, z_cart)
+    elevation = N.arctan2(z_cart, hypot_xy)
+    azimuth = N.arctan2(y_cart, x_cart)
+    return azimuth, elevation, radius
+
+
+def getscangeoloc(lat_start, lon_start, lat_end, lon_end):
```
Comment (skymandr) [inline app/zpt/handler/geoloc_tools.py:20]:
```suggestion
def get_scan_geoloc(lat_start, lon_start, lat_end, lon_end):
```
or maybe even `...geo_loc` – very hard to read function name!

File: app/zpt/handler/geos.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import numpy as np
+
+
+def geoid_radius(latitude):
```
Comment (skymandr) [inline app/zpt/handler/geos.py:4]: It feels like a bad idea to implement this ourselves. Is this even allowed by Craig Haley's license whatever it is..?

File: app/zpt/handler/msis90.py [hunk 6]
```diff
@@ -21,66 +17,67 @@
...
-            d = np.zeros(9, 'f')
-            t = np.zeros(2, 'f')
-            nrlmsis.gtd7(
-                iydd, ut, alt, lat, lng, ut / 3600 - lng / 15, f107a, f107, ap,
-                mass, d, t,
+            dens, temp = msise_model(
+                datetime, alt, lat, lng, f107a, f107, apavg, flags=[1] * 24
             )
+            t = np.array(temp)
+            d = np.array(dens)
```
Comment (skymandr) [inline app/zpt/handler/msis90.py:56]: Couldn't you make sure that `msise_model` returns arrays instead? Or is this from the external model?

File: app/zpt/handler/msis90.py [hunk 7]
```diff
@@ -21,66 +17,67 @@
...
+                datetime,
+                alt,
+                lat,
+                lng,
+                f107a,
+                f107,
+                apavg,
             )
+            t = np.array(temp)
+            d = np.array(dens)
```
Comment (skymandr) [inline app/zpt/handler/msis90.py:80]: same here

File: app/zpt/handler/newdonalettyERANC.py [hunk 8]
```diff
@@ -1,339 +1,71 @@
...
-AVOGADRO = 6.02282e23  # [mol^-1] aovogadros number
 Ro = 8.3143  # [J * mol^-1 * K^-1] ideal gas constant
 
 
 class Donaletty:
     """
-         Create ZPT2 file using the new NetCDF ECMWF files
-         Created
-             Donal Murtagh July 2011.
+    Create ZPT2 file using the new NetCDF ECMWF files
```
Comment (skymandr) [inline app/zpt/handler/newdonalettyERANC.py:13]: `PTZ2`?

File: app/zpt/handler/newdonalettyERANC.py [hunk 9]
```diff
@@ -1,339 +1,71 @@
...
         """Inputs :
-                    g5zpt : heights (km) Pressure (hPa) and temperature profile
-                            should extend to at least 60 km
-                    scan_datetime : datetime object
-                    lat : latitude of observation
-            Output : ZPT array [3, 0-120]
+                g5zpt : heights (km) Pressure (hPa) and temperature profile
+                        should extend to at least 60 km
+                scan_datetime : datetime object
+                lat : latitude of observation
```
Comment (skymandr) [inline app/zpt/handler/newdonalettyERANC.py:24]: No longer true

File: app/zpt/handler/newdonalettyERANC.py [hunk 10]
```diff
@@ -1,339 +1,71 @@
...
-            '{prefix}_pl_{date}-{hour}.nc'.format(
-                prefix=prefix,
-                date=date.strftime('%Y-%m-%d'),
-                hour=hour
+    def makeprofile(self, scans: xarray.Dataset):
+        ecmz = np.arange(45)
+        newz = np.arange(151)
+        scan_on_interp_gmh = scans.groupby("scanid").map(
+            lambda ds: ds.swap_dims(level="era5_gmh").interp(
+                era5_gmh=ecmz, kwargs={"fill_value": 273}
```
Comment (skymandr) [inline app/zpt/handler/newdonalettyERANC.py:65]: Magic number. Probably 0 ˚C/237.15 K?

File: app/zpt/handler/scan_data_descriptions.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
+parameter_desc = {
```
Comment (skymandr) [inline app/zpt/handler/scan_data_descriptions.py:1]: Neat!

File: app/zpt/requirements.in [hunk 12]
```diff
@@ -1,4 +1,13 @@
 -c ${CONSTRAINTS}
+arrow
```
Comment (skymandr) [inline app/zpt/requirements.in:2]: Do we actually need `arrow`?

**Review by skymandr — COMMENTED** [review]
Overall looks like an improvement. I have a few comments, though I realise they are probably "old sins" and not yours, though you find yourself in the somewhat unfortunate situation of having had them land in your lap, as it were.

Regarding the constants: changing those may cause some very small discrepancies with what was produced before, but I think using the defined values is better than trying to keep track of what old textbook each value used comes from.

I also note that Boltzmann's constant is defined in at least two places and with different values.

**Comment by joakimmoller** [comment]
This is implemented in level0

**Pull request closed without merging**